Civi/Core/SettingsManager.php

```diff
@@ -328,38 +394,56 @@ public function flush() {
   }
 
   /**
-   * Get a list of critical system defaults.
-   *
-   * The setting system can be modified by extensions, which means that it's not fully available
-   * during bootstrap -- in particular, defaults cannot be loaded. For a very small number of settings,
-   * we must define defaults before the system bootstraps.
-   *
-   * @param string $entity
+   * Load boot settings from environment and/or settings file
    *
-   * @return array
+   * @param string $settingsPath
+   *    Path to the civicrm.settings.php file
    */
-  private static function getSystemDefaults($entity) {
-    $defaults = [];
-    switch ($entity) {
-      case 'domain':
-        $defaults = [
-          'installed' => FALSE,
-          // The default list of components should be kept in sync with "civicrm_extension.sqldata.php".
-          'enable_components' => ['CiviEvent', 'CiviContribute', 'CiviMember', 'CiviMail', 'CiviReport', 'CiviPledge'],
-          'customFileUploadDir' => '[civicrm.files]/custom/',
-          'imageUploadDir' => '[civicrm.files]/persist/contribute/',
-          'uploadDir' => '[civicrm.files]/upload/',
-          'imageUploadURL' => '[civicrm.files]/persist/contribute/',
-          'extensionsDir' => '[civicrm.files]/ext/',
-          'extensionsURL' => '[civicrm.files]/ext/',
-          'ext_max_depth' => \CRM_Extension_System::DEFAULT_MAX_DEPTH,
-          'resourceBase' => '[civicrm.root]/',
-          'userFrameworkResourceURL' => '[civicrm.root]/',
-        ];
-        break;
+  public static function bootSettings($settingsPath) {
+    // during bootstrap we can't use a more persistent cache
+    $bootSettingsCache = \CRM_Utils_Cache::create([
+      'name' => 'bootSettings',
+      'type' => ['ArrayCache'],
+    ]);
+
+    $bootSettingsManager = new self($bootSettingsCache);
+
+    // check for constants we need to define
+    $bootConstants = SettingsMetadata::getMetadata(['is_contact' => 0, 'is_constant' => TRUE], NULL, FALSE, TRUE);
+
+    // if a constant value has been set using an env var, we need
+    // to jump in and define it now so the env var value takes precedence
+    // over any "define" calls in the civicrm.settings.php
+    // (hopefully these use if (!defined(X))))
+    $envSettingsValues = $bootSettingsManager->getEnvSettingValues('domain');
+
+    foreach ($bootConstants as $key => $meta) {
+      $fqn = $meta['global_name'] ?? NULL;
+      $value = $envSettingsValues[$key] ?? NULL;
+      if ($fqn && !defined($fqn) && !is_null($value)) {
+        define($fqn, $value);
+      }
+    }
 
+    if (file_exists($settingsPath)) {
+      if (!defined('CIVICRM_SETTINGS_PATH')) {
+        define('CIVICRM_SETTINGS_PATH', $settingsPath);
+        require_once $settingsPath;
+      }
```

Comment:
@ufundo  When I do `civibuild reinstall` on my `standalone-clean` site, I get a crash. The main install step runs:

```bash
cv core:install ...
```

But then calls to `cv` fail:

```bash
cv ev 'return "hello world";'

cv api3 system.get

cv ext:list -L

cv ev 'echo CIVICRM_SETTINGS_PATH;'
## ^^ This is actually the next step of `civibuild reinstall`, but it's coincidental.
```

Specifically, the commands fail at:

```
[PHP Warning] require_once(/CRM/Core/ClassLoader.php): failed to open stream: No such file or directory at phar:///Users/totten/bknix/bin/cv/lib/src/Bootstrap.php:93

Fatal error: require_once(): Failed opening required '/CRM/Core/ClassLoader.php' (include_path='.:/Users/totten/bknix/build/s3/web/core:/Users/totten/bknix/build/s3/web/core/packages:/Users/totten/bknix/build/s3/web/core/vendor/tecnickcom:/Users/totten/bknix/build/s3/web/core/vendor/pear/auth_sasl:/Users/totten/bknix/build/s3/web/core/vendor/pear/console_getopt:/Users/totten/bknix/build/s3/web/core/vendor/pear/db:/Users/totten/bknix/build/s3/web/core/vendor/pear/log:/Users/totten/bknix/build/s3/web/core/vendor/pear/mail:/Users/totten/bknix/build/s3/web/core/vendor/pear/mail_mime:/Users/totten/bknix/build/s3/web/core/vendor/pear/net_smtp:/Users/totten/bknix/build/s3/web/core/vendor/pear/net_socket:/Users/totten/bknix/build/s3/web/core/vendor/pear/pear-core-minimal/src:/Users/totten/bknix/build/s3/web/core/vendor/pear/pear_exception:/Users/totten/bknix/build/s3/web/core/vendor/pear/validate_finance_creditcard:.:/nix/store/sx8q6l82wbdzhkqbyapka4f5pwncijwl-php-7.4.29/lib/php') in phar:///Users/totten/bknix/bin/cv/lib/src/Bootstrap.php on line 93
```

The problem?

* `$civicrm_root` is undefined.... because `civicrm.settings.php` hasn't loaded at all. 
* Recall https://github.com/civicrm/cv/pull/193/commits/3837693b1d7e27eb95d0e598461e3523c8b4c028 -- it changed `findCivicrmSettingsPhp()` to treat `civicrm.standalone.php` as a work-a-like for `civicrm.settings.php`.
*  As a consequence, `Bootstrap.php` puts `/path/to/my/civicrm.standalone.php` [into `CIVICRM_SETTINGS_PATH`](https://github.com/civicrm/cv/blob/v0.3.51/lib/src/Bootstrap.php#L211).
* Now this code (`SettingsManager::bootSettings()`) considers loading the real `civicrm.settings.php`... and declines.
* So we never load the real `civicrm.settings.php`.

Why did it work before? Because it didn't matter if `CIVICRM_SETTINGS_PATH` was slightly munged --  [it proceeded to load anyway](https://github.com/civicrm/civicrm-core/blob/5.76/setup/res/civicrm.standalone.php.txt#L28).

My first resolution was https://gist.github.com/totten/5d71c4b4a01e2f1678b3929265c5ab67, but that feels like putting a hack on top of a hack. I'm not sure pretty way to do that.

Perhaps this would be cleaner: make `SettingsManager::bootSettings()` behave more like the prior revision of `civicrm.standalone.php.txt`? i.e.

```diff
diff --git a/Civi/Core/SettingsManager.php b/Civi/Core/SettingsManager.php
index 54b1146a27..5e92566e16 100644
--- a/Civi/Core/SettingsManager.php
+++ b/Civi/Core/SettingsManager.php
@@ -428,8 +428,8 @@ class SettingsManager {
     if (file_exists($settingsPath)) {
       if (!defined('CIVICRM_SETTINGS_PATH')) {
         define('CIVICRM_SETTINGS_PATH', $settingsPath);
-        require_once $settingsPath;
       }
+      require_once $settingsPath;
     }

     // get all effective values from the settings bag (resolving defaults etc)
```